lib/ansible/plugins/action/gather_facts.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import time
+
+from ansible import constants as C
+from ansible.plugins.action import ActionBase
+from ansible.utils.vars import combine_vars
+
+
+class ActionModule(ActionBase):
+
+    def _get_module_args(self, fact_module, task_vars):
+
+        mod_args = self._task.args.copy()
+
+        # deal with 'setup specific arguments'
+        if fact_module != 'setup':
```

Comment:
@bcoca Is there a reason why only `all` is allowed for subset?

I added this to make it work for other subset values as well in networking.
```
         # deal with 'setup specific arguments'
         if fact_module != 'setup':
-            subset = mod_args.pop('gather_subset', None)
+            connection = task_vars.get('ansible_connection')
 
-            if subset not in ('all', ['all']):
-                self._display.warning('Ignoring subset(%s) for %s' % (subset, fact_module))
+            if connection not in ('network_cli', 'httpapi', 'netconf'):
+                subset = mod_args.pop('gather_subset', None)
+                if subset not in ('all', ['all']):
+                    self._display.warning('Ignoring subset(%s) for %s' % (subset, fact_module))
```

Reply:
subset is a parameter of `setup` and not supported by many other modules, we can make an exception list based on module, no need to do it on connection basis, specially if you end up with same logic

Reply:
@bcoca this is mostly for all the facts module we have for networking.

Reply:
@trishnaguha i think a better fix for networking would be to: 
 - change the default setting to 'smart'
 - on 'smart' it should examine connection and use either 'mapped' module or fallback to setup.py

Reply:
@bcoca with these lines of code, I was trying to make `gather_subset` work with network platforms.
For example playbook being
```
- hosts: eos
  connection: network_cli
  gather_facts: yes
  gather_subset: config

  vars_files:
    - group_vars/eos.yaml

  tasks:
  - debug:
      var: ansible_facts
```
output would be
```
ok: [veos01] => {
    "ansible_facts": {
        "_facts_gathered": true, 
        "net_config": "! Command: show running-config\n! device: an-veos-01 (vEOS, EOS-4.15.9M)\n!\n! boot system flash:EOS.swi\n!\ntransceiver qsfp default-mode 4x10G\n!\nhostname an-veos-01\nip name-server vrf default 10.8.36.1\nip name-server vrf default 10.8.38.1\nip domain-name ansible.eng.rdu2.redhat.com\n!\nspanning-tree mode mstp\n!\nno aaa root\n!\nusername admin privilege 15 role network-admin secret 5 $1$oNbWHCQz$c.hX28usAI2jrY9llfWda1\nusername ansible privilege 15 role network-admin secret 5 $1$FPjf6c7h$2hnf6B43SUrFQcDwviLod0\nusername bashuser nopassword\n!\nvlan 2-9,12\n!\nvlan 10\n   name vlan_10\n!\nvlan 11\n   name vlan_11\n!\nvrf definition test\n!\ninterface Ethernet1\n   no switchport\n!\ninterface Ethernet2\n   no switchport\n!\ninterface Ethernet3\n!\ninterface Ethernet4\n!\ninterface Ethernet5\n!\ninterface Ethernet6\n!\ninterface Ethernet7\n!\ninterface Management1\n   ip address 10.8.38.31/24\n!\nip route 10.0.0.0/8 10.8.38.1\n!\nip routing\nno ip routing vrf test\n!\nrouter bgp 10\n!\nmanagement api http-commands\n   protocol http\n   protocol unix-socket\n   no shutdown\n!\n!\nend", 
        "net_fqdn": "an-veos-01.ansible.eng.rdu2.redhat.com", 
        "net_gather_subset": [
            "default", 
            "config"
        ], 
        "net_hostname": "an-veos-01", 
        "net_image": "flash:EOS.swi", 
        "net_model": "vEOS", 
        "net_serialnum": "", 
        "net_version": "4.15.9M"
    }
}
```
It would just gather the `config` subset rather than `all` for the EOS platform in this playbook.

Reply:
ah, for some reason i was under the opposite impression, that those modules did NOT support gather_subset

Reply:
@trishnaguha so our networking facts modules do support gather_subset, but doing this by connection means that 3rd party ones MUST also support it to be used in this fashion

Reply:
@bcoca Not sure if you have noticed, the prefix `ansible_` is stripped out from every keys in ansible_facts dictionary. But if I run the *_facts module itself, the keys have `ansible_` prefix.

Reply:
@bcoca we can provide warning in that case and default to all?

Reply:
@bcoca All the network facts modules have `gather_subset` arg being default to some value.

Reply:
i know, i checked, it is just an observation for the case where a user wants to use a custom module